clients/src/main/java/org/apache/kafka/common/network/DefaultChannelMetadataRegistry.java

```diff
@@ -16,15 +16,15 @@
  */
 package org.apache.kafka.common.network;
 
+import java.util.Objects;
+
 public class DefaultChannelMetadataRegistry implements ChannelMetadataRegistry {
     private CipherInformation cipherInformation;
     private ClientInformation clientInformation;
 
     @Override
     public void registerCipherInformation(final CipherInformation cipherInformation) {
-        if (this.cipherInformation != null) {
-            this.cipherInformation = cipherInformation;
-        }
+        this.cipherInformation =  Objects.requireNonNull(cipherInformation);
```

Comment:
Could you elaborate on what is wrong with the previous implementation? If there is a bug, would it be possible to reproduce it in a unit test?

Reply:
+1 to add a unit test.

The previous implementation will not update the cipherInformation if one already exists. This is inconsistent with the implementation in other implementation of `ChannelMetadataRegistry` i.e. SelectorChannelMetadataRegistry, hence, makes me believe that the expected behaviour is to replace the member if it already exists. Note that this file is only used in tests so far.

Reply:
I don't recall why we added this check. I think that the cipher can only be set once per connection so this may be the reason. I suppose that it is fine to remove it in this case. I am not sure about the `Objects.requireNonNull` though.

As this file is only used in tests, it would also make sense to move it to the `test` source tree.

Reply:
Thanks @dajac @divijvaidya for the review. I've added a new test following this change.

Reply:
>  I am not sure about the Objects.requireNonNull though.

can you expand on this @dajac ?

> As this file is only used in tests, it would also make sense to move it to the test source tree.

I am okay with this. Will we want to keep the added unit test in this case, any thoughts? @dajac @divijvaidya

Reply:
ping @dajac @divijvaidya

Reply:
> Will we want to keep the added unit test in this case, any thoughts?

No. unit test for a test will not be required in that case. Let's move that file to tests. Although it's a public class but it cannot be used outside of Kafka code base.